x/paychan/model.go

```diff
@@ -64,10 +69,26 @@ func (pc PaymentChannel) Copy() orm.CloneableData {
 
 // NewPaymentChannelBucket returns a bucket for storing PaymentChannel state.
 func NewPaymentChannelBucket() orm.ModelBucket {
-	b := orm.NewModelBucket("paychan", &PaymentChannel{})
+	b := orm.NewModelBucket("paychan", &PaymentChannel{},
+		orm.WithIDSequence(paymentChannelSeq))
 	return migration.NewModelBucket("paychan", b)
 }
 
+// Declare it globally so that it can be reference by both the bucket and the
+// peekNextID function.
+var paymentChannelSeq = orm.NewSequence("paychan", "id")
+
+// peekNextID returns the next ID that will be used by the payment channel ID
+// sequence. This function is not thread safe and relies on the sequence state
+// in the database.
+func peekNextID(db weave.KVStore) (int64, error) {
```

Comment:
I kind of like this approach, because it is explicit.
I would love to see it adopted in other extensions too.

If you don't like the idea of adopting buckets/interfaces - you can create a global helper method to peek and make sure it knows how to peek in each bucket implementation. WDYT?

Reply:
I think I don't fully understand your motivation yet. Right now I have quite the opposite opinion on what is a better approach.

I think we both agree that we are comparing two solutions to create the address that requires the entity ID. 

1. A solution which is embedded in the creation as done in `aswap` package
https://github.com/iov-one/weave/blob/d43242272a704ce8fe845845578643a532f55095/x/aswap/model.go#L105-L111

2. A solution which inspects what is the next sequence value in order to guess the next ID
https://github.com/iov-one/weave/blob/d43242272a704ce8fe845845578643a532f55095/x/paychan/handler.go#L66-L75

In both cases, to compute the entity address I need its ID (`key`). What we seem to differ in opinion is the way that entity ID is obtained. 
By now we ruled out the solution with save & update. This is a good choice as otherwise entity `Validate` method could not fail on empty `Address` attribute.

In solution 1. an "old" way of implementing a bucket is used. `aswap.Bucket` inherits from the `orm.Bucket` and extends it with additional functionality. Additional functionality is the [sequence generation](https://github.com/iov-one/weave/blob/d43242272a704ce8fe845845578643a532f55095/x/aswap/model.go#L83) hidden behind the `Create` method.
`Create` method is a perfect place to compute created entity address, because this is where we have access to the key. This is also the place that stores the entity. 
The only weak point of this solution in my opinion is that the original entity is mutated and `.Address` attribute is set.

In solution 2. a `orm.ModelBucket` is used. This implementation does not allow for inheritance. It also hides from the user [the ID generation](https://github.com/iov-one/weave/blob/d43242272a704ce8fe845845578643a532f55095/orm/model_bucket.go#L53-L57).
I do not want to grow the `orm.ModelBucket` interface to leak out the implementation. My solution was to request the bucket to use [my sequence instance](https://github.com/iov-one/weave/blob/d43242272a704ce8fe845845578643a532f55095/x/paychan/model.go#L73), so that I can access its state in database independently of the bucket. Knowing where the sequence keeps its state, I can read it and guess the next value.
This solution is dangerous as you can be wrong guessing the value. You do not acquire and own the ID value like in the previous example. You only guess it and trust that it will be given to you next time you ask the sequence to grant you an ID.

I cannot see why [explicit managing the sequence in the `Create` method](https://github.com/iov-one/weave/blob/d43242272a704ce8fe845845578643a532f55095/x/aswap/model.go#L105-L116) is magical. On the other hand, relying on a global sequence instance and database state not being modified, guessing the value you maybe will own in the future is explicit.
Of the above do not make an argument then even comparing the amount of code required to implement this makes it hard for me to think that the 2. is any better.

I just came up with a better solution for the second case. I will implement it shortly.

Reply:
I have updated solution 2. and I think now it is good enough and comparable to 1.

The difference with the previous version is that [I am explicitly acquiring the ID](https://github.com/iov-one/weave/blob/b9cf347e61ff7ca2aebea14de14f6994e11c4715/x/paychan/handler.go#L72-L77) which I am later using to insert the entity for the first time. This is a safe implementation because I own the ID returned by the sequence. It is also shorter and requires less explanation. Only for safety and "just in case" I am requesting the bucket to [use the same sequence instance](https://github.com/iov-one/weave/blob/b9cf347e61ff7ca2aebea14de14f6994e11c4715/x/paychan/model.go#L72).

Reply:
> This solution is dangerous as you can be wrong guessing the value. You do not acquire and own the ID value like in the previous example. You only guess it and trust that it will be given to you next time you ask the sequence to grant you an ID.

Since transactions and messages they bear are executed sequentially in weave by nature - this approach seems to be safe enough. The implementation is another question. 



> The only weak point of this solution in my opinion is that the original entity is mutated and `.Address` attribute is set.

This is exactly what I'm talking about, I didn't choose the words correctly on that one. The issue here is that we go and assign all the fields we need in the handler code and then mutate the object after. Hence my preference for peek, which makes the whole process of assembling the object more transparent.

Reply:
> `Create` method is a perfect place to compute created entity address, because this is where we have access to the key. This is also the place that stores the entity.

It is indeed a perfect place for this code, but in my opinion `Create` should not have any side-effects in terms of object contents.

Reply:
Makes sense. Now I think I understand your motivation better.

What do you think about the update that I have described in https://github.com/iov-one/weave/pull/891#discussion_r304271708?

Reply:
@husio I have finally found the original thread about side-effects in `aswap` implementation: https://github.com/iov-one/weave/pull/571#discussion_r281772914, just in case the reasoning there is a bit more clear than what I'm trying to express. 

I still think we should utilise exactly the same approach across our extensions and preferably the most transparent one.

Reply:
> Makes sense. Now I think I understand your motivation better.
> 
> What do you think about the update that I have described in [#891 (comment)](https://github.com/iov-one/weave/pull/891#discussion_r304271708)?

I like this approach, I think I have already commented on it. 
It does not speculate, but instead uses the same id it's peeking to actually save the data.